### DISCUSSION TITLE
ARM64 Build

### DISCUSSION THREAD

**Discussion by choff5507** [qpdf/qpdf#925] [Q&A]
I am hoping someone can help me get a build that works on ARM. Seems like the ones included in the repos for ARM are v7 or v8 which seems quite old. Trying to build my own from source Ive ran into some issues and there doesn't seem to be a binary version posted with the normal releases here. 

I did notice that the ArchLinux repo seemed to have the most current version but simply downloading it and then trying to run it gives an error. 

Anyone have any advice on how I can get a current working version that can be used on AARCH64?

Thanks!

**Comment by jberkenbilt** [comment]
It would help if you provided some information about what is not working. qpdf is routinely built on numerous platforms, so I'd be surprised if it isn't working. Most likely it's a toolchain issue. Sometimes a newer compiler or older than qpdf is built with causes a problem. If you're having a problem on a particular distribution, you might reach out to whoever packages it for that distribution. I know the latest version builds for arm64 on debian with no issues.

**Reply by choff5507** [reply]
Thanks for the reply. Im trying to build on Oracle Linux. Ive pasted the error below. Im sure it's probably something Im missing as Im far from an expert. I did make sure that `libjeg` is installed. 



`-- Using -Wall: YES
-- Using -Wconversion: YES
-- Using -Wsign-conversion: YES
-- Using -Wshadow=local: YES
-- Using -Wold-style-cast: YES
-- Checking for module 'libjpeg'
--   Package 'libjpeg', required by 'virtual:world', not found
CMake Error at libqpdf/CMakeLists.txt:188 (message):
  libjpeg not found


-- Checking for module 'gnutls'
--   Package 'gnutls', required by 'virtual:world', not found
CMake Error at libqpdf/CMakeLists.txt:278 (message):
  Missing dependencies; unable to continue


-- Configuring incomplete, errors occurred!
See also "/home/opc/qpdf_src/qpdf-11.3.0/build/CMakeFiles/CMakeOutput.log".
See also "/home/opc/qpdf_src/qpdf-11.3.0/build/CMakeFiles/CMakeError.log".
`

**Reply by jberkenbilt** [reply]
This is pretty much a basic build issue and doesn't have anything to do with qpdf. You need to have qpdf's dependencies installed in order to build it, so you will need to install development libraries for libjpeg and gnutls. I would direct you here: https://qpdf.readthedocs.io/en/stable/installation.html plus to general knowledge about to build things from source.

**Comment by Haribo112** [comment]
We build our ARM64 image using Docker:

```
FROM debian:buster-slim AS qpdf-build

RUN apt-get update -yqq; apt-get install -y --no-install-recommends \
      build-essential \
      ca-certificates \
      libssl-dev \
      libjpeg-dev \
      zlib1g \
      zlib1g-dev \
      wget && \
    wget -q https://github.com/qpdf/qpdf/releases/download/v11.3.0/qpdf-11.3.0.tar.gz && \
    tar -xzf qpdf-11.3.0.tar.gz 

RUN wget -q https://github.com/Kitware/CMake/releases/download/v3.26.0/cmake-3.26.0-linux-$(uname -m | sed 's/arm64/aarch64/g').sh -O cmake-3.26.0-linux.sh && \
    chmod +x cmake-3.26.0-linux.sh && \
    mkdir /cmake && \
    ./cmake-3.26.0-linux.sh --skip-license --prefix=/cmake && \
    ln -s /cmake/bin/cmake /usr/bin/cmake

WORKDIR /qpdf-11.3.0

RUN cmake -S . -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo && \
    cmake --build build && cmake --install build && ldconfig
